zipkin-lens/src/components/TracePage/TraceTimeline/TraceTree.jsx

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright 2015-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import React from 'react';
+import { withStyles } from '@material-ui/styles';
+import grey from '@material-ui/core/colors/grey';
+
+import {
+  spanBarLinePosY,
+  spanToggleButtonTranslate,
+  spanTreeLineWidthPercentPerDepth,
+  spanTreeWidthPercent,
+  spanToggleButtonLengthOfSide,
+} from '../sizing';
+
+const style = theme => ({
+  line: {
+    stroke: theme.palette.grey[500],
+    strokeWidth: '1px',
+  },
+  spanToggleButton: {
+    cursor: 'pointer',
+    opacity: 0,
+    backgroundColor: grey[300],
+    '&:hover': {
+      opacity: 0.1,
+    },
+  },
+  buttonIcon: {
+    stroke: theme.palette.grey[800],
+    strokeWidth: 2,
+  },
+  spanToggleButtonInternal: {
+    stroke: theme.palette.grey[700],
+    strokeWidth: 1,
+    fill: theme.palette.common.white,
+  },
+});
+
+const TraceTree = ({
+  spans,
+  depth,
+  closedSpans,
+  onSpanToggleButtonClick,
+  classes,
+}) => {
```

Comment:
At a glance it feels like this component is doing quite many things here.. but can't be sure until measure it :D

Reply:
But I do think these logics should be extract to out of this function/component. They are not really simple.

Reply:
How often this component will re-render? I think [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo) can be quite useful here to increase rendering performance (just in case if it is slow -- otherwise we will be only doing premature optimization)